sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala

```diff
@@ -115,26 +116,45 @@ class ResolveHiveSerdeTable(session: SparkSession) extends Rule[LogicalPlan] {
 
 class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case filterPlan @ Filter(_, SubqueryAlias(_, relation: HiveTableRelation)) =>
+      val predicates = PhysicalOperation.unapply(filterPlan).map(_._2).getOrElse(Nil)
+      computeTableStats(relation, predicates)
     case relation: HiveTableRelation
         if DDLUtils.isHiveTable(relation.tableMeta) && relation.tableMeta.stats.isEmpty =>
-      val table = relation.tableMeta
-      val sizeInBytes = if (session.sessionState.conf.fallBackToHdfsForStatsEnabled) {
-        try {
-          val hadoopConf = session.sessionState.newHadoopConf()
-          val tablePath = new Path(table.location)
-          val fs: FileSystem = tablePath.getFileSystem(hadoopConf)
-          fs.getContentSummary(tablePath).getLength
-        } catch {
-          case e: IOException =>
-            logWarning("Failed to get table size from hdfs.", e)
-            session.sessionState.conf.defaultSizeInBytes
-        }
-      } else {
-        session.sessionState.conf.defaultSizeInBytes
+      computeTableStats(relation)
+  }
+
+  private def computeTableStats(
+      relation: HiveTableRelation,
+      predicates: Seq[Expression] = Nil): LogicalPlan = {
+    val table = relation.tableMeta
+    val sizeInBytes = if (session.sessionState.conf.fallBackToHdfsForStatsEnabled) {
+      try {
+        val hadoopConf = session.sessionState.newHadoopConf()
+        val tablePath = new Path(table.location)
+        val fs: FileSystem = tablePath.getFileSystem(hadoopConf)
+        BigInt(fs.getContentSummary(tablePath).getLength)
+      } catch {
+        case e: IOException =>
+          logWarning("Failed to get table size from hdfs.", e)
+          getSizeInBytesFromTablePartitions(table.identifier, predicates)
       }
+    } else {
+      getSizeInBytesFromTablePartitions(table.identifier, predicates)
+    }
+    val withStats = table.copy(stats = Some(CatalogStatistics(sizeInBytes = sizeInBytes)))
+    relation.copy(tableMeta = withStats)
+  }
 
-      val withStats = table.copy(stats = Some(CatalogStatistics(sizeInBytes = BigInt(sizeInBytes))))
-      relation.copy(tableMeta = withStats)
+  private def getSizeInBytesFromTablePartitions(
+      tableIdentifier: TableIdentifier,
+      predicates: Seq[Expression] = Nil): BigInt = {
+    session.sessionState.catalog.listPartitionsByFilter(tableIdentifier, predicates) match {
```

Comment:
Have you tested the performance of `session.sessionState.catalog.listPartitionsByFilter(tableIdentifier, predicates)` and `fs.getContentSummary(tablePath).getLength`?

Reply:
The perf will be pretty bad when the number of partitions is huge. Thus, I think we can close this PR.

Reply:
After [this refactor](https://github.com/apache/spark/pull/22743). We can avoid compute stats if `LogicalRelation` already cached. because the computed stats will not take effect.

Reply:
Your solution seems much better.  @wangyum 
Thank you.  I'll close it.

Reply:
How come https://github.com/apache/spark/pull/22743 solves this problem? That PR targets to invalidate cache when configurations are changed. This PR targets to compute stats from HDFS when they are not available.